integration-tests/src/tests/client/features/nonrefundable_transfer.rs

```diff
@@ -0,0 +1,346 @@
+        .nightshade_runtimes(&genesis)
+        .build()
+}
+
+/// Creates a test environment using default protocol version.
+fn setup_env() -> TestEnv {
+    setup_env_with_protocol_version(None)
+}
+
+fn get_nonce(env: &mut TestEnv, signer: &InMemorySigner) -> u64 {
+    let request = QueryRequest::ViewAccessKey {
+        account_id: signer.account_id.clone(),
+        public_key: signer.public_key.clone(),
+    };
+    match env.query_view(request).unwrap().kind {
+        QueryResponseKind::AccessKey(view) => view.nonce,
+        _ => panic!("wrong query response"),
+    }
+}
+
+fn account_exists(env: &mut TestEnv, account_id: AccountId) -> bool {
+    let request = QueryRequest::ViewAccount { account_id };
+    env.query_view(request).is_ok()
+}
+
+fn execute_transaction_from_actions(
+    env: &mut TestEnv,
+    actions: Vec<Action>,
+    signer: &InMemorySigner,
+    receiver: AccountId,
+) -> Result<FinalExecutionOutcomeView, InvalidTxError> {
+    let tip = env.clients[0].chain.head().unwrap();
+    let nonce = get_nonce(env, signer);
+    let tx = SignedTransaction::from_actions(
+        nonce + 1,
+        signer.account_id.clone(),
+        receiver,
+        signer,
+        actions,
+        tip.last_block_hash,
+    );
+    let tx_result = env.execute_tx(tx);
+    let height = env.clients[0].chain.head().unwrap().height;
+    for i in 0..2 {
+        env.produce_block(0, height + 1 + i);
+    }
+    tx_result
+}
+
+/// Submits a transfer (either regular or non-refundable).
+///
+/// This methods checks that the balance is subtracted from the sender and added
+/// to the receiver, if the status was ok. No checks are done on an error.
+fn exec_transfer(
+    env: &mut TestEnv,
+    signer: InMemorySigner,
+    receiver: AccountId,
+    deposit: Balance,
+    nonrefundable: bool,
+    account_creation: bool,
+    implicit_account_creation: bool,
+    deploy_contract: bool,
+) -> Result<FinalExecutionOutcomeView, InvalidTxError> {
+    let sender_pre_balance = env.query_balance(sender());
+    let (receiver_before_amount, receiver_before_nonrefundable) = if account_creation {
+        (0, 0)
+    } else {
+        let receiver_before = env.query_account(receiver.clone());
+        (receiver_before.amount, receiver_before.nonrefundable)
+    };
+
+    let mut actions = vec![];
+
+    if account_creation && !implicit_account_creation {
+        actions.push(Action::CreateAccount(CreateAccountAction {}));
+        actions.push(Action::AddKey(Box::new(AddKeyAction {
+            public_key: PublicKey::from_seed(KeyType::ED25519, receiver.as_str()),
+            access_key: AccessKey { nonce: 0, permission: AccessKeyPermission::FullAccess },
+        })));
+    }
+
+    if nonrefundable {
+        actions.push(Action::NonrefundableStorageTransfer(NonrefundableStorageTransferAction {
+            deposit,
+        }));
+    } else {
+        actions.push(Action::Transfer(TransferAction { deposit }));
+    }
```

Comment:
I've modified tests so that 3 options are tested:
* `NonrefundableStorageTransfer` only
* both kind of transfers
  * non-refundable transfer first
  * regular transfer first